ibtk/src/utilities/IBTK_MPI.cpp

```diff
@@ -0,0 +1,334 @@
+} // getCommunicator
+
+IBTK_MPI::comm
+IBTK_MPI::getSAMRAIWorld()
+{
+#if SAMRAI_VERSION_MAJOR == 2
+    return SAMRAI_MPI::commWorld;
+#else
+    return SAMRAI_MPI::getSAMRAIWorld()
+#endif
+}
+
+int
+IBTK_MPI::getNodes(IBTK_MPI::comm communicator /* = MPI_COMM_NULL */)
+{
+    communicator = communicator == MPI_COMM_NULL ? d_communicator : communicator;
+    int nodes = 1;
+    MPI_Comm_size(communicator, &nodes);
+    return nodes;
+} // getNodes
+
+int
+IBTK_MPI::getRank(IBTK_MPI::comm communicator /* = MPI_COMM_NULL */)
+{
+    communicator = communicator == MPI_COMM_NULL ? d_communicator : communicator;
+    int node = 0;
+    MPI_Comm_rank(communicator, &node);
+    return node;
+} // getRank
+
+void
+IBTK_MPI::barrier(IBTK_MPI::comm communicator /* = MPI_COMM_NULL */)
+{
+    communicator = communicator == MPI_COMM_NULL ? d_communicator : communicator;
+    (void)MPI_Barrier(communicator);
+} // barrier
+
+template <typename T>
+T
+IBTK_MPI::minReduction(T x, int* rank_of_min, IBTK_MPI::comm communicator)
+{
+    minReduction(&x, 1, rank_of_min, communicator);
+    return x;
+} // minReduction
+
+template <typename T>
+void
+IBTK_MPI::minReduction(T* x, const int n, int* rank_of_min, IBTK_MPI::comm communicator)
+{
+    communicator = communicator == MPI_COMM_NULL ? d_communicator : communicator;
+    if (n == 0) return;
+    if (rank_of_min == nullptr)
+    {
+        std::vector<T> send(n);
+        std::copy(x, x + n, send.begin());
+        MPI_Allreduce(send.data(), x, n, mpi_type_id(send[0]), MPI_MIN, communicator);
+    }
+    else
+    {
+        minMaxReduction(x, n, rank_of_min, MPI_MINLOC, communicator);
+    }
+} // minReduction
+
+template <typename T>
+T
+IBTK_MPI::maxReduction(T x, int* rank_of_max, IBTK_MPI::comm communicator)
+{
+    maxReduction(&x, 1, rank_of_max, communicator);
+    return x;
+} // maxReduction
+
+template <typename T>
+void
+IBTK_MPI::maxReduction(T* x, const int n, int* rank_of_max, IBTK_MPI::comm communicator)
+{
+    communicator = communicator == MPI_COMM_NULL ? d_communicator : communicator;
+    if (n == 0) return;
+    if (rank_of_max == nullptr)
+    {
+        std::vector<T> send(n);
+        std::copy(x, x + n, send.begin());
+        MPI_Allreduce(send.data(), x, n, mpi_type_id(send[0]), MPI_MAX, communicator);
+    }
+    else
+    {
+        minMaxReduction(x, n, rank_of_max, MPI_MAXLOC, communicator);
+    }
+} // maxReduction
+
+template <typename T>
+T
+IBTK_MPI::sumReduction(T x, IBTK_MPI::comm communicator)
+{
+    sumReduction(&x, 1, communicator);
+    return x;
+} // sumReduction
+
+template <typename T>
+void
+IBTK_MPI::sumReduction(T* x, const int n, IBTK_MPI::comm communicator)
+{
+    communicator = communicator == MPI_COMM_NULL ? d_communicator : communicator;
+    if (n == 0 || getNodes(communicator) < 2) return;
+    std::vector<T> send(n);
+    std::copy(x, x + n, send.begin());
+    MPI_Allreduce(send.data(), x, n, mpi_type_id(send[0]), MPI_SUM, communicator);
```

Comment:
We can skip the copy by setting the output buffer to `MPI_IN_PLACE`. That may not have been around when the SAMRAI guys wrote this.